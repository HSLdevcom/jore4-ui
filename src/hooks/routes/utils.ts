export const extractFirstAndLastStopFromStops = (
  stopIdsWithinRoute: UUID[],
) => {
  // TODO: These will be removed from schema, remove from here as well
  const startingStopId = stopIdsWithinRoute[0];
  const finalStopId = stopIdsWithinRoute[stopIdsWithinRoute.length - 1];

  return { startingStopId, finalStopId };
};
