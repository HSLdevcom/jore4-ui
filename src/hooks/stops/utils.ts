import uniqBy from 'lodash/uniqBy';
import {
  JourneyPatternJourneyPattern,
  ServicePatternScheduledStopPoint,
} from '../../generated/graphql';

// checking whether this stop is the start or end stop of an existing route
export const isStartingOrEndingStopOfAnyRoute = (
  stopWithRouteGraphData: ServicePatternScheduledStopPoint,
) => {
  const stopId = stopWithRouteGraphData.scheduled_stop_point_id;
  return stopWithRouteGraphData.scheduled_stop_point_in_journey_patterns.some(
    (item) => {
      // journey patterns/routes that this stop is part of
      const route = item.journey_pattern.journey_pattern_route;
      // is the stop the start or end stop of this given route
      return (
        stopId === route?.starts_from_scheduled_stop_point_id ||
        stopId === route?.ends_at_scheduled_stop_point_id
      );
    },
  );
};

// gets the unique list of parent routes for the input journey patterns
export const getRoutesOfJourneyPatterns = (
  journeyPatterns: JourneyPatternJourneyPattern[],
) => {
  const allRoutes = journeyPatterns
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    .map((item) => item.journey_pattern_route!);

  // in the future, multiple journey patterns may have the same route,
  // so let's make sure we only return unique results
  return uniqBy(allRoutes, (route) => route.route_id);
};
