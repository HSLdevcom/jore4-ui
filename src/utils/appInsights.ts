import { ReactPlugin } from '@microsoft/applicationinsights-react-js';
import { ApplicationInsights } from '@microsoft/applicationinsights-web';
import { createBrowserHistory } from 'history';

const browserHistory = createBrowserHistory();
export const reactPlugin = new ReactPlugin();
export const appInsights = new ApplicationInsights({
  config: {
    instrumentationKey: 'YOUR_INSTRUMENTATION_KEY_GOES_HERE',
    extensions: [reactPlugin],
    extensionConfig: {
      [reactPlugin.identifier]: { history: browserHistory },
    },
  },
});
appInsights.loadAppInsights();
