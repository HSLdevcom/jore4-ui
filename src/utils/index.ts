export * from './forms';
export * from './url';
