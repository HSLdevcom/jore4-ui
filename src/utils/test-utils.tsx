// Based on https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react'; // eslint-disable-line import/no-extraneous-dependencies
import React, { FunctionComponent } from 'react';

const AllTheProviders: FunctionComponent = ({ children }) => {
  // Add "providers" or "wrappers" that are needed in all DOM render tests here
  return <>{children}</>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';
// override render method
export { customRender as render };
