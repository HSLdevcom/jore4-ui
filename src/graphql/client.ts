import {
  ApolloClient,
  concat,
  HttpLink,
  InMemoryCache,
  split,
} from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { getMainDefinition } from '@apollo/client/utilities';
import { mapHttpToWs } from '../utils';
import { authRoleMiddleware } from './auth';

export const createGraphqlClient = () => {
  const graphqlServerUrl = '/api/graphql/v1/graphql';

  const httpLink = new HttpLink({
    uri: graphqlServerUrl,
  });

  const apolloLink = concat(authRoleMiddleware, httpLink);

  // because next.js might run this on server-side and websockets aren't
  // supported there, we have to check if we are on browser before
  // initializing WebSocket link
  const wsLink = process.browser
    ? new WebSocketLink({
        // WebSocketLink doesn't work with relative url's, so we have to
        // turn relative url into absolute.
        uri: `${mapHttpToWs(window.location.origin)}${graphqlServerUrl}`,
        options: {
          reconnect: true,
        },
      })
    : undefined;

  const link = process.browser
    ? split(
        ({ query }) => {
          const definition = getMainDefinition(query);
          const isSubscription =
            definition.kind === 'OperationDefinition' &&
            definition.operation === 'subscription';
          return isSubscription;
        },
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        wsLink!,
        apolloLink,
      )
    : apolloLink;

  const cache = new InMemoryCache({
    typePolicies: {
      route_route: {
        keyFields: ['route_id'],
      },
      service_pattern_scheduled_stop_point: {
        keyFields: ['scheduled_stop_point_id'],
      },
    },
  });

  const client = new ApolloClient({
    link,
    cache,
    defaultOptions: {
      watchQuery: {
        // Should return cached query results first while rerunning the query against the backend.
        // After the backend response has returned, the query results are updated
        fetchPolicy: 'cache-and-network',
      },
    },
  });

  return client;
};
