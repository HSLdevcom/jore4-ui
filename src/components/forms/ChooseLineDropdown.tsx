import React from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine, useListAllLinesQuery } from '../../generated/graphql';
import {
  FormInputProps as ListboxInputProps,
  Listbox,
} from '../../uiComponents';

interface Props extends ListboxInputProps {
  testId: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <span>{`${item.name_i18n} (${item.description_i18n})`}</span>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { data: lines } = useListAllLinesQuery();
  const { t } = useTranslation();

  const options = lines?.route_line.map(mapToOption) || [];

  const selectedLine = lines?.route_line.find((item) => item.line_id === value);

  // if no line is selected, show "Choose line"
  const mapToButtonContent = (displayedLine?: RouteLine) => (
    <div className="w-full">
      {displayedLine
        ? `${displayedLine.name_i18n} (${displayedLine.description_i18n})`
        : t('routes.chooseLine')}
    </div>
  );

  return (
    <Listbox
      testId={testId}
      buttonContent={mapToButtonContent(selectedLine)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};
