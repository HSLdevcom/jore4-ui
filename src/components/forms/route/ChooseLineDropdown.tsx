import React from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine, useListAllLinesQuery } from '../../../generated/graphql';
import {
  FormInputProps as ListboxInputProps,
  Listbox,
} from '../../../uiComponents';

interface Props extends ListboxInputProps {
  testId?: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <span>{`${item.label} (${item.name_i18n})`}</span>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { data: lines } = useListAllLinesQuery();
  const { t } = useTranslation();

  // @ts-expect-error typings are off?
  const options = lines?.route_line.map(mapToOption) || [];

  const selectedLine = lines?.route_line.find(
    (item) => item.line_id === value,
  ) as RouteLine;

  // if no line is selected, show "Choose line"
  const mapToButtonContent = (displayedLine?: RouteLine) => (
    <div className="w-full">
      {displayedLine
        ? `${displayedLine.label} (${displayedLine.name_i18n})`
        : t('routes.chooseLine')}
    </div>
  );

  return (
    <Listbox
      testId={testId}
      buttonContent={mapToButtonContent(selectedLine)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};
