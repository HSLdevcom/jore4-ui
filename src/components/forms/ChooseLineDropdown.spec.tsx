import { MockedProvider } from '@apollo/client/testing';
import { act, fireEvent, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import React from 'react';
import {
  ListAllLinesDocument,
  useListAllLinesQuery,
} from '../../generated/graphql';
import { render, sleep } from '../../utils/test-utils';
import { ChooseLineDropdown } from './ChooseLineDropdown';

describe('<ChooseLineDropdown />', () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onChange = () => {};
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onBlur = () => {};
  const value = 'line1';
  const testId = 'chooseLineDropdown1';

  const mocks = [
    {
      request: {
        query: ListAllLinesDocument,
        variables: {},
      },
      result: {
        data: {
          route_line: [
            {
              __typename: 'route_line',
              line_id: 'line1',
              label: '1',
              name_i18n: 'Line1 name',
              short_name_i18n: 'Line1',
              validity_start: DateTime.fromISO('2017-02-13T12:51:48.000Z'),
              validity_end: null,
            },
            {
              __typename: 'route_line',
              line_id: 'line2',
              label: '2',
              name_i18n: 'Line2 name',
              short_name_i18n: 'Line2',
              validity_start: DateTime.fromISO('2017-02-13T12:51:48.000Z'),
              validity_end: null,
            },
          ],
        },
      } as ReturnType<typeof useListAllLinesQuery>,
    },
  ];

  test('Opens dropdown when clicked and shows all lines', async () => {
    const { container, asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseLineDropdown
          testId={testId}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
        />
      </MockedProvider>,
    );

    // dropdown is collapsed, value is not yet bound to dropdown, 'Choose line' text shows
    expect(screen.getByTestId(testId)).toHaveTextContent('routes.chooseLine');
    expect(container.querySelector('li')).toBeNull();
    expect(asFragment()).toMatchSnapshot();

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // dropdown is collapsed, data is loaded and the pre-selected line1 is shown on the button
    expect(screen.getByTestId(testId)).toHaveTextContent('1 (Line1 name)');
    expect(container.querySelector('li')).toBeNull();
    expect(asFragment()).toMatchSnapshot();

    // click dropdown to open it:
    const openDropdownButton = screen.getByTestId(testId);
    fireEvent.click(openDropdownButton);

    // dropdown is open, both lines show
    expect(screen.queryByText('routes.chooseLine')).toBeNull();
    const items = container.querySelectorAll('li');
    expect(items[0].textContent).toBe('1 (Line1 name)');
    expect(items[1].textContent).toBe('2 (Line2 name)');
    expect(asFragment()).toMatchSnapshot();
  });
});
