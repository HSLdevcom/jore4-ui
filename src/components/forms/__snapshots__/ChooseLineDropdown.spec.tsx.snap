// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseLineDropdown /> Opens dropdown when clicked and shows all lines 1`] = `
<DocumentFragment>
  <div
    class="input-element relative h-full"
    id="listbox"
  >
    <button
      aria-haspopup="true"
      class="flex h-full w-full items-center text-left focus:outline-none"
      data-testid="chooseLineDropdown1"
      id="headlessui-listbox-button-1"
      type="button"
    >
      <div
        class="w-full"
      >
        routes.chooseLine
      </div>
      <i
        class="icon-arrow ml-auto text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Opens dropdown when clicked and shows all lines 2`] = `
<DocumentFragment>
  <div
    class="input-element relative h-full"
    id="listbox"
  >
    <button
      aria-haspopup="true"
      class="flex h-full w-full items-center text-left focus:outline-none"
      data-testid="chooseLineDropdown1"
      id="headlessui-listbox-button-1"
      type="button"
    >
      <div
        class="w-full"
      >
        1 (Line1 name)
      </div>
      <i
        class="icon-arrow ml-auto text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Opens dropdown when clicked and shows all lines 3`] = `
<DocumentFragment>
  <div
    class="input-element relative h-full"
    id="listbox"
  >
    <button
      aria-controls="headlessui-listbox-options-3"
      aria-expanded="true"
      aria-haspopup="true"
      class="flex h-full w-full items-center text-left focus:outline-none"
      data-testid="chooseLineDropdown1"
      id="headlessui-listbox-button-1"
      type="button"
    >
      <div
        class="w-full"
      >
        1 (Line1 name)
      </div>
      <i
        class="icon-arrow ml-auto text-tweaked-brand transition duration-150 ease-in-out -rotate-180"
        style="font-size: 10px;"
      />
    </button>
    <ul
      aria-activedescendant="headlessui-listbox-option-4"
      aria-labelledby="headlessui-listbox-button-1"
      class="absolute left-0 mt-2 w-full rounded-b-md border border-black border-opacity-20 bg-white shadow-md focus:outline-none transition ease-out duration-100 opacity-0 scale-95"
      id="headlessui-listbox-options-3"
      role="listbox"
      tabindex="0"
    >
      <div>
        <li
          aria-selected="true"
          id="headlessui-listbox-option-4"
          role="option"
          tabindex="-1"
        >
          <span
            class="bg-background flex group text-left px-2 py-2 focus:outline-none"
          >
            1 (Line1 name)
          </span>
        </li>
        <li
          id="headlessui-listbox-option-5"
          role="option"
          tabindex="-1"
        >
          <span
            class=" flex group text-left px-2 py-2 focus:outline-none"
          >
            2 (Line2 name)
          </span>
        </li>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
